vllm/v1/core/kv_cache_manager.py

```diff
@@ -368,10 +387,80 @@ def get_num_common_prefix_blocks(
                 break
         return num_common_blocks
 
+    def _free_block_hashes_by_request_id(self, request_id: str) -> None:
+        """Free the block hashes allocated for the request.
+
+        Args:
+            request_id: The request ID to free the block hashes.
+        """
+        self.req_to_block_hashes.pop(request_id, None)
+
     def free_block_hashes(self, request: Request) -> None:
         """Discard the block hashes for the request.
 
         NOTE: Unlike `free`, this method should be called only when the request
         is finished, not when it is preempted.
         """
-        self.req_to_block_hashes.pop(request.request_id, None)
+        self._free_block_hashes_by_request_id(request.request_id)
+
+    def alloc_and_get_external_blocks(
+        self,
+        request: "Request",
+        computed_blocks: list["KVCacheBlock"],
+        num_computed_tokens: int,
+        kv_connector: "KVConnectorBase_V1",
+    ) -> tuple[list["KVCacheBlock"], int]:
+
+        # Check for cache hit.
+        need_to_allocate = kv_connector.get_num_matched_tokens(
+            request, num_computed_tokens)
+        num_allocated_blocks = 0
+
+        # Cache hit: allocate buffer.
+        if need_to_allocate > 0:
+            # HACK: We don't want the scheduler see the blocks are allocated
+            # and associated with the current request. Instead, we want the
+            # scheduler find that the blocks are already allocated and they
+            # are associated with some other requests (i.e., the case of
+            # prefix caching.
+
+            old_req_id = request.request_id
+            request.request_id = f"{old_req_id}-buf-for-kv-connector"
+            allocated_blocks = self.allocate_slots(
```

Comment:
As this call replace the `request_id` to the new one, so reference to https://github.com/vllm-project/vllm/pull/15960/files#diff-e3795fc9494f612a6504d432ba9e01b7ed640f0aa73c6db8cb9be56357f3e751R291 , the `req_to_block_hashes` is always `0` for caller `alloc_and_get_external_blocks`, but the `num_cached_blocks` is not `0` while prefix cache hit, so the `assert len(block_hashes) >= num_cached_blocks` is not make sense for this scenario.

@ApostaC @robertgshaw2-redhat @zhoufang5-tal @captainzmc what do you think? This is a blocker for reuse purpose.

Reply:
@zhoufang5-tal  I think this is the root cause of your issue, I have encountered the same issue with you.

I just remove the `assert` and go through the code, it can workaround.

Reply:
This case actually is not related to prefix caching. The case where this can occur is when rescheduling a preempted request that has a remote KV

I added an assert for this case. We can consider it more closely in the follow up PR.